### PR TITLE
T1. Fix isolated connection bugs, improve tests, upgrade dependencies

### LINKED ISSUE #1687: connect_isolated only works on mainnet
**Is your feature request related to a problem? Please describe.**

`zebra_network::isolated::connect_isolated` uses the default `zebra_network::Config`, which is configured for mainnet.

But the Zcash network protocol uses different magic numbers for mainnet and testnet in its message headers, so `connect_isolated` will fail on testnet:
https://github.com/ZcashFoundation/zebra/blob/69ba5584f34f5af81bc381494ef39a7a2c11da75/zebra-network/src/protocol/external/codec.rs#L161

**Describe the solution you'd like**

- [ ] Add a `Network` argument to `connect_isolated`, and substitute it for the default network in `peer::Handshake::builder`. The easiest way to do this might be to add a `with_network` method to `peer::Handshake::builder`.
- [ ] Add tests for the `connect_isolated` handshake on mainnet and testnet.

There's no need to actually send a transaction as part of the tests - a successful handshake or an informational query should be enough.

**Additional context**

This change isn't required until we need to send isolated user-generated transactions from `zebra-client`.

### LINKED ISSUE #2999: Remove outdated http dependencies
## Motivation

Zebra currently depends on 2 different versions of the `http` crate.
One of these versions is over a year old.

If we remove outdated dependencies, Zebra will compile faster, its binaries will be smaller, and it will be more secure.

## Tasks

0. Check that the conflict still exists by removing the `http` exception from `deny.toml`, then running `cargo deny check bans`.

Upgrade versions:
1. Fork `h2`, upgrade it to use the latest `http`, and make Zebra depend on our forked versions
2. Submit PRs upstream
3. Remove the exceptions for these crates in `deny.toml`.

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

This PR contains some `zebra-network` changes that are required or useful for Tor support using `arti`. But it doesn't add any `arti` dependencies or Tor code yet. (That's in the next PR.)

We should merge this PR because these are all improvements, and we will have to do them eventually anyway.

It also makes testing easier, because it lets us write tests that don't depend on the network. And it improves test coverage.

## Solution

Enhancements:
- Make handshakes generic over AsyncRead + AsyncWrite
- Add a wrapper function for isolated TCP connections to an IP address

Dependencies:
- Upgrade Zebra dependencies to be compatible with arti, but don't add any Tor dependencies yet
    - This just happens to close #2999

Fixes & Cleanups:
- Simplify connect_isolated using ServiceExt::map_err and BoxError
- Support connect_isolated on testnet, closes #1687 
- Move isolated network tests to their own module

Testing:
- Improve isolated TCP connection tests
- Add an in-memory connection test that uses AsyncReadWrite

## Review

This PR is ready for review, @dconnolly can review it.

It might have merge conflicts with future PRs, so we should get it in soon.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Actually add Tor support.

### CODE REVIEW

**Comment by teor2345** [comment]
This PR conflicts with PR #3292 or the #3295 series of PRs, so I'm marking it as draft until they merge.

File: zebra-network/src/isolated.rs [hunk 1]
```diff
@@ -1,62 +1,66 @@
...
-/// The connection pool returned by `init` should be used for all requests that
+/// The connection pool returned by [`init`](zebra_network::init)
+/// should be used for all requests that
 /// don't require isolated state or use of an existing TCP connection. However,
 /// this low-level API is useful for custom network crawlers or Tor connections.
 ///
 /// In addition to being completely isolated from all other node state, this
 /// method also aims to be minimally distinguishable from other clients.
 ///
+/// SECURITY TODO: check if the timestamp field can be zeroed, to remove another distinguisher (#3300)
```
Comment (dconnolly) [inline zebra-network/src/isolated.rs:34]: 👍

**Comment by conradoplg** [comment]
I've synced with main and solved conflicts

**Review by conradoplg — APPROVED** [review]
Looks good!

**Pull request merged**